feature(transpiler): add key to transpile method

jacksonopp opened the pull request:
## PR Checklist

Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/ngneat/transloco/blob/master/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
  - I may need help with testing since this feature doesn't add any functionality directly, but it does not break any existing tests afaik
- [x] Docs have been added / updated (for bug fixes / features)

## PR Type

What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->

```
[ ] Bugfix
[x] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Documentation content changes
[ ] Other... Please describe:
```

## What is the current behavior?

<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->
https://github.com/ngneat/transloco/issues/568

The `TranslocoTranspiler.transpile` does not give the key of the current translation.

Issue Number: 568

## What is the new behavior?

The `TranslocoTranspiler.transpile` now accepts the key as a parameter, and the `TranslocoService.translate` method now passes in the key.

## Does this PR introduce a breaking change?

```
[ ] Yes
[x] No
```

<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. -->

## Other information
This is one of my first contributions to an open source project, so I expect that I will need to make some changes.

shaharkazaz commented on package.json:
```diff
@@ -123,4 +123,3 @@
     }
   }
 }
-
```
Please revert the changes in the package and package lock.

jacksonopp replied:
This has been reverted

shaharkazaz commented on docs/yarn.lock:
```diff
@@ -2681,7 +2681,7 @@ class-utils@^0.3.5:
     isobject "^3.0.0"
```
Please revert.

jacksonopp replied:
This has been reverted

shaharkazaz replied:
Seems like it's still here

jacksonopp replied:
🤦 I'll fix it again

jacksonopp replied:
This should be fixed again. 

Thank you for your patience as I work thru my first OSS contribution by the way!

shaharkazaz commented on docs/docs/transpiler.mdx:
```diff
@@ -112,7 +112,7 @@ You can also provide a custom transpiler by creating a class that implements the
 
 export class CustomTranspiler implements TranslocoTranspiler {
-  transpile(value: any, params, translation: Translation) {
+  transpile(value: any, params, translation: Translation, key?: any) {
```
Why is this optional and typed as any?

jacksonopp replied:
Updated to non-optional. 

More discussion here: https://github.com/ngneat/transloco/pull/569#discussion_r887043400

shaharkazaz commented on apps/transloco-playground/src/assets/i18n/custom-transpilers/es.json:
```diff
@@ -0,0 +1,5 @@
+{
```
I don't see the point in adding this to the docs, this is a very specific implementation you needed.

shaharkazaz replied:
I think you can make a simpler example by providing a custom transpiler that logs the paramaters.

shaharkazaz commented on libs/transloco/src/lib/transloco.transpiler.ts:
```diff
@@ -16,7 +16,7 @@ import {
 
 export interface TranslocoTranspiler {
-  transpile(value: any, params: HashMap, translation: Translation): any;
+  transpile(value: any, params: HashMap, translation: Translation, key?: any): any;
```
Why is this optional and typed as any? Please add the following comment above:
```
// TODO Change parameters to object in the next major release
```

jacksonopp replied:
Added TODO.

I added `key` as optional since the built in transpilers don't consume it, and it allowed me to change less of the core library. However after thinking about it more I understand better how the transpilers are hooked up, so I have removed the optional type, and am now passing in 'key' wherever `transpile` is being called.

This does mean that I had to update the specs to now pass in a key. I have not added any specs since because this isn't really adding any functionality to the built in transpilers, i'm not sure what exactly I should be testing. I will definitely need help understanding what I should be looking for in the specs.

As for the `any` type, I was following the type of the `value` parameter. Looking at the `translocoService.translate` method, `key` should be `TranslateParams` correct?

shaharkazaz replied:
@jacksonopp I don't think we need to add any specs here, it's just passing another param.
The key type is `string`

shaharkazaz commented:
@jacksonopp fix the types please and force push all the changes in one commit made using `yarn commit`/`npm rnu commit`.

shaharkazaz commented on docs/docs/plugins/message-format.mdx:
```diff
@@ -18,7 +18,7 @@ npm i messageformat @ngneat/transloco-messageformat
 The `MessageFormatTranspiler` is compatible with the `DefaultTranspiler` and therefore you can switch without worry that it will break your current translations.
 
-It then enables support for the following within youIt enables support for the following within your i18n translation files:
+It then enables support for the following within your i18n translation files:
```
Are you updated with master? this was already fixed

jacksonopp replied:
This should be resolved now

shaharkazaz commented on apps/transloco-playground/src/app/custom-transpilers/custom-transpilers.component.html:
```diff
@@ -0,0 +1,17 @@
+<h2>Custom Transpilers</h2>
+<p>In order to test this, you will need to provide the <code>customTranspiler</code> in app.module.ts</p>
```
Provide the custom transpiler in the app module, it's just logging the params now 🙂 
I don't think we need a dedicated page.

jacksonopp replied:
Removed page

shaharkazaz commented on apps/transloco-playground/src/app/custom-transpilers/custom.transpiler.ts:
```diff
@@ -0,0 +1,13 @@
+import { HashMap, Translation, TranslocoTranspiler, TRANSLOCO_TRANSPILER } from "@ngneat/transloco";
+
+export class CustomTranspiler implements TranslocoTranspiler {
+  transpile(value: any, params: HashMap<any>, translation: Translation, key: any) {
+    console.log({value, params, translation, key})
+    return value
+  }
+}
```
```suggestion
export class CustomTranspiler implements TranslocoTranspiler {
  transpile(value: any, params: HashMap<any>, translation: Translation, key: any) {
    console.log('Custom Transpiler called with:', {value, params, translation, key});
    
    return value;
  }
}
```

jacksonopp replied:
Good suggestion, I have made this change

shaharkazaz commented on libs/transloco/src/lib/transloco.transpiler.ts:
```diff
@@ -81,7 +82,8 @@ export class DefaultTranspiler implements TranslocoTranspiler {
     params: HashMap = {},
-    translation: Translation
+    translation: Translation,
+    key: any
```
why any?

jacksonopp replied:
Changed to `string`

shaharkazaz commented on libs/transloco/src/lib/transloco.transpiler.ts:
```diff
@@ -104,9 +106,10 @@ export class DefaultTranspiler implements TranslocoTranspiler {
     params: HashMap = {},
-    translation: Translation
+    translation: Translation,
+    key: any
```
why any?

jacksonopp replied:
Changed to `string`

shaharkazaz commented on libs/transloco/src/lib/transloco.transpiler.ts:
```diff
@@ -147,7 +150,7 @@ export class FunctionalTranspiler
   }
 
-  transpile(value: any, params: HashMap = {}, translation: Translation): any {
+  transpile(value: any, params: HashMap = {}, translation: Translation, key: any): any {
```
why any?

jacksonopp replied:
Changed to `string`

jacksonopp commented:
Thank you for your patience as I work thru my first OSS contribution by the way!

shaharkazaz commented on apps/transloco-playground/src/app/custom.transpiler.ts:
```diff
@@ -0,0 +1,13 @@
+import { HashMap, Translation, TranslocoTranspiler, TRANSLOCO_TRANSPILER } from "@ngneat/transloco";
+
+export class CustomTranspiler implements TranslocoTranspiler {
+  transpile(value: any, params: HashMap<any>, translation: Translation, key: any) {
```
```suggestion
  transpile(value: any, params: HashMap<any>, translation: Translation, key: string) {
```

jacksonopp replied:
Fixed

shaharkazaz commented:
@jacksonopp please note that the CI is failing

jacksonopp commented:
@shaharkazaz 

### Looking into the unit tests
Looks like we're seeing a peer dependency mismatch around rxjs. I think running an npm install should fix this, and would involve committing a change to package-lock.json. Is that okay with you?

### Looking into the e2e tests, I'm seeing the following failures:

- `locale.spec.ts` is failing because it is displaying the wrong unix date, and it fails regardless of whether or not we are using the custom transpiler. I was playing around with changing my time zone, and that seemed to affect the displayed output. Adding `timeZone: 'UTC'` as a parameter to the `translocoDate` pipe normalizes the timezone, and seems to solve this issue. I'm not sure what has changed to cause this. Does this sound correct to you?

- `full-cycle.spec.ts` is failing because the custom transpiler just returns the value without doing any transpiling. Removing the custom transpiler from the providers array seems to fix this issue. Since we've updated the documentation site with the new `key` input, I think its fine if we do not include this in the transloco playground. It wasn't there before anyway.

Thoughts?

shaharkazaz commented:
@jacksonopp 
* You didn't make any change on the package so I'm not sure why would that change anything 🤔 But let's try it, make sure you have node 16+ and npm 8+.
* Let's try to remove the custom compiler from the playground.
* I don't think there was any change in the locale area, and last couple of commits passed, I'll take a look when I get time if the issue persists, anyways you didn't cause the issue 🙂

jacksonopp commented:
@shaharkazaz 

- Updated package-lock.json. I used `node v16.15.1` and `npm 8.11.0`
- Removed the custom transpiler from the playground
- Running through the specs, I'm now seeing the unit tests around locales failing for the same reason as the e2e tests, but sounds like you're going to investigate it.

jacksonopp commented:
@shaharkazaz Any update on this? Looks like all specs are passing now.

shaharkazaz commented:
@jacksonopp sorry man I'm swamped, I'll get to it as soon as I have some free time to do one last review 🙏

jacksonopp commented:
@shaharkazaz I understand you are probably busy, but any more updates?

shaharkazaz commented:
@jacksonopp Sorry man for disappearing, had a lot of stuff going on and just couldn't find the time.
If you'd like we can pick this up again.

jacksonopp commented:
@shaharkazaz All good. I've updated my fork to match master here.

shaharkazaz commented:
@jacksonopp nice work on this man. I hope to see you contributing more 😄
